internal/stanza/converter.go

```diff
@@ -15,49 +15,276 @@
+	"go.uber.org/zap"
 )
 
-// Convert a stanza-style entry to a pdata.Logs
-func Convert(obsLog *entry.Entry) pdata.Logs {
+// Converter converts entry.Entry into pdata.Logs aggregating translated
+// entries into logs coming from the same Resource.
+// Logs are being sent out based on the flush interval and/or the maximum
+// batch size.
+type Converter struct {
+	// pLogsChan is a channel on which batched logs will be sent to.
+	pLogsChan chan pdata.Logs
+
+	stopOnce sync.Once
+	stopChan chan struct{}
+
+	// flushInterval defines how often we flush the aggregated log entries.
+	flushInterval time.Duration
+	// flushTriggerCount defines what's the amount of entries in the buffer that
+	// will trigger a flush of log entries.
+	flushTriggerCount uint
+	// flushChan is an internal signalling channel to trigger flushing of
+	// log entries.
+	flushChan chan struct{}
+
+	// data is the internal cache which is flushed regularly, either when
+	// flushInterval ticker ticks or when max number of entries for a
+	// particular Resource is reached.
+	data      map[string][]*entry.Entry
+	dataMutex sync.RWMutex
+	dataCount uint
+
+	wg sync.WaitGroup
+
+	logger *zap.Logger
+}
+
+type ConverterOption interface {
+	apply(*Converter)
+}
+
+type optionFunc func(*Converter)
+
+func (f optionFunc) apply(c *Converter) {
+	f(c)
+}
+
+func WithFlushInterval(interval time.Duration) ConverterOption {
+	return optionFunc(func(c *Converter) {
+		c.flushInterval = interval
+	})
+}
+
+func WithFlushTriggerCount(count uint) ConverterOption {
+	return optionFunc(func(c *Converter) {
+		c.flushTriggerCount = count
+	})
+}
+
+func WithLogger(logger *zap.Logger) ConverterOption {
+	return optionFunc(func(c *Converter) {
+		c.logger = logger
+	})
+}
+
+const (
+	defaultFlushInterval = time.Second
+	defaultFlushMaxSize  = 100
+)
+
+func NewConverter(opts ...ConverterOption) *Converter {
+	c := &Converter{
+		pLogsChan:         make(chan pdata.Logs),
+		stopChan:          make(chan struct{}),
+		logger:            zap.NewNop(),
+		flushChan:         make(chan struct{}, 1),
+		flushInterval:     defaultFlushInterval,
+		flushTriggerCount: defaultFlushMaxSize,
+		data:              make(map[string][]*entry.Entry),
+		wg:                sync.WaitGroup{},
+	}
+
+	for _, opt := range opts {
+		opt.apply(c)
+	}
+
+	return c
+}
+
+func (c *Converter) Start() {
+	c.logger.Debug("Starting log converter")
+
+	c.wg.Add(1)
+	go func(c *Converter) {
+		defer c.wg.Done()
+
+		ticker := time.NewTicker(c.flushInterval)
+		defer ticker.Stop()
+
+		flush := func() {
+			if err := c.flush(context.Background()); err != nil {
+				c.logger.Error("Problem sending log entries",
+					zap.Error(err),
+				)
+			}
+		}
+		for {
+			select {
+			case <-c.stopChan:
+				return
+
+			case <-ticker.C:
+				flush()
+
+			case <-c.flushChan:
+				flush()
+				// NOTE:
+				// Since we've received a flush signal independently of flush
+				// ticker do we want to reset the flush ticker?
```

Comment:
OK cool. So I had the `ticker.Reset()` call here initially but then I realized that it was introduced in 1.15 and the CI is [based on 1.14](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/fcbe766a2bb18f52a0e0b8b07b8a1d074a264044/.circleci/config.yml).

Should we update the CI as well?

Reply:
This won't be up to me. 1.14 is also set in the root `go.mod` so I'm presuming that we should work within 1.14 for now.

I suspect there is another reasonable way to achieve the same behavior? If not, I believe it would be reasonable to capture the desired behavior in a new issue and leave as is for now.

Reply:
I'll create an issue if/when this PR gets merged so that I can reference it properly (without hypothesizing in the issue description)